File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-2.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        ROW, COL = len(matrix), len(matrix[0])
        dp = [[-1] * COL for _ in range(ROW)]
        res = 1
        dirs = [(-1, 0), (1, 0), (0, 1), (0, -1)]

        def dfs(r, c):
            if dp[r][c] > -1:
                return dp[r][c]
            temp = []
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if nr < 0 or nc < 0 or nr >= ROW or nc >= COL or matrix[nr][nc] >= matrix[r][c] :
                    continue
                temp.append(dfs(nr, nc))
            if not temp:
                dp[r][c] = 1
            else:
                dp[r][c] = max(temp) + 1
            return dp[r][c]
        
        mv = min(min(row) for row in matrix)
        for r in range(ROW):
            for c in range(COL):
                if matrix[r][c] == mv:
                    dp[r][c] = 1

        for r in range(ROW):
            for c in range(COL):
                if dp[r][c] == -1:
                    dfs(r, c)
                res = max(dp[r][c], res)
        return res
                

File: Data Structures & Algorithms/trapping-rain-water/submission-2.cpp
class Solution {
public:
    int trap(vector<int>& height) {
       int n = height.size();
       int res = 0;
       vector<int> left_max(n, 0);
       vector<int> right_max(n, 0);
       for (int i = 1; i < n; ++i) {
            left_max[i] = max(left_max[i - 1], height[i - 1]);
            right_max[n - i - 1] = max(right_max[n - i], height[n - i]);
       }
        for (int i = 1; i < n - 1; ++i) {
            res += max(0, min(left_max[i], right_max[i]) - height[i]);
        }
        return res;
    }
};
